source-code/core/src/lint/createLintRule.ts

```diff
@@ -0,0 +1,56 @@
+import type { LintRule } from "./rule.js"
+
+/**
+ * A utility function to create a lint rule.
+ *
+ * @example
+ * const myRule = createLintRule({ id: "example.rule" }, (args) => {
+ *   // Implement your rule logic here
+ * });
+ *
+ * // using the rule
+ * myRule("error");
+ *
+ * // you can also pass settings to the rule
+ * const myRule2 = createLintRule<{ strict: boolean }>({ id: "example.rule" }, (args) => {
+ *   // Implement your rule logic here
+ * });
+ *
+ * // settings must be provided when using the rule
+ * myRule2("error", { strict: true });
+ */
+export const createLintRule = <Settings extends object | undefined = undefined>(
```

Comment:
no more required and optional Settings?

Reply:
Still exists with easier types 😎

Checkout the type tests https://github.com/inlang/inlang/pull/459/files#diff-d3442843c87c4325671801e6262b8ce1e656aef697b1b87f6238968d1f026144. 

The type that provides optional types is straightforward

```
/**
 * Type for the function that configures a lint rule.
 */
type ConfigureLintRuleFunction<Settings extends Record<string, unknown> | undefined = undefined> =
	Settings extends undefined
		? // If settings are not defined, the function only takes a level
		  (level: LintRule["level"]) => LintRule
		: // Else, the function takes a level and settings
		  (level: LintRule["level"], settings: Settings) => LintRule
```

Reply:
Not exactly true..
```ts
const rule = createLintRule<{ strict?: boolean }>()

// this shows an error
rule('error')
// you need to pass an empty object
rule('error', { })
```

Reply:
I'll merge this for now. To unblock the editor team and @jannesblobel. We can add a boolean flag later. I vote against it though. Makes the lint usage more complicated and passing an empty object for defined settings with all optional properties is minor IMHO.